pkg/planner/core/plan_cost_ver2.go

```diff
@@ -127,20 +136,38 @@ func (p *PhysicalTableScan) GetPlanCostVer2(taskType property.TaskType, option *
 	}
 
 	rows := getCardinality(p, option.CostFlag)
-	var rowSize float64
-	if p.StoreType == kv.TiKV {
-		rowSize = getAvgRowSize(p.StatsInfo(), p.tblCols) // consider all columns if TiKV
+
+	var columns []*expression.Column
+	if p.StoreType == kv.TiKV { // Assume all columns for TiKV
+		columns = p.tblCols
 	} else { // TiFlash
-		rowSize = getAvgRowSize(p.StatsInfo(), p.schema.Columns)
+		columns = p.schema.Columns
 	}
-	rowSize = math.Max(rowSize, 2.0)
-	scanFactor := getTaskScanFactorVer2(p, p.StoreType, taskType)
+	rowSize := getAvgRowSize(p.StatsInfo(), columns)
+	// Ensure rowSize has a reasonable minimum value to avoid underestimation
+	rowSize = math.Max(rowSize, MinRowSize)
 
+	scanFactor := getTaskScanFactorVer2(p, p.StoreType, taskType)
 	p.PlanCostVer2 = scanCostVer2(option, rows, rowSize, scanFactor)
 
-	// give TiFlash a start-up cost to let the optimizer prefers to use TiKV to process small table scans.
+	// Apply TiFlash startup cost to prefer TiKV for small table scans
 	if p.StoreType == kv.TiFlash {
-		p.PlanCostVer2 = costusage.SumCostVer2(p.PlanCostVer2, scanCostVer2(option, 10000, rowSize, scanFactor))
+		p.PlanCostVer2 = costusage.SumCostVer2(p.PlanCostVer2, scanCostVer2(option, TiFlashStartupRowPenalty, rowSize, scanFactor))
+	} else {
+		// Apply cost penalty for high modifyCount or if PreferRangeScan enabled and pseudo/zero stats
+		sessionVars := p.SCtx().GetSessionVars()
+		allowPreferRangeScan := sessionVars.GetAllowPreferRangeScan()
+		tblColHists := p.tblColHists
+
+		preferRangeScanCondition := allowPreferRangeScan && (tblColHists.Pseudo || tblColHists.RealtimeCount < 1)
+		hasHighModifyCount := tblColHists.ModifyCount > tblColHists.RealtimeCount
+		hasLowEstimate := rows > 1 && int64(rows) < tblColHists.RealtimeCount && int64(rows) <= tblColHists.ModifyCount
+		shouldApplyPenalty := !p.isChildOfIndexLookUp && (preferRangeScanCondition || hasHighModifyCount || hasLowEstimate)
+
+		if shouldApplyPenalty {
```

Comment:
Better to add some comments for this, and add https://github.com/pingcap/tidb/issues/56012 as an unit test.